File: tools/release_helper/summary.py
"""
Summary generation utilities for the release helper.
"""

import json
from tools.release_helper.metadata import list_all_apps
from tools.release_helper.validation import is_prerelease_version


def generate_release_summary(
    matrix_json: str,
    version: str,
    event_type: str,
    dry_run: bool = False,
    repository_owner: str = ""
) -> str:
    """Generate a release summary for GitHub Actions."""

    try:
        matrix = json.loads(matrix_json) if matrix_json else {"include": []}
    except json.JSONDecodeError:
        matrix = {"include": []}

    summary = []
    summary.append("## 🚀 Release Summary")
    summary.append("")
    
    if not matrix.get("include"):
        summary.append("🔍 **Result:** No apps detected for release")
    else:
        summary.append("✅ **Result:** Release completed")
        summary.append("")
        
        apps = [item["app"] for item in matrix["include"]]
        summary.append(f"📦 **Apps:** {', '.join(apps)}")
        
        # Handle version display - show individual versions if they differ from the main version
        app_versions = [item.get("version", version) for item in matrix["include"]]
        unique_versions = list(set(app_versions))
        
        # Check if any version is a prerelease
        has_prerelease = any(is_prerelease_version(v) for v in app_versions if v)
        
        if len(unique_versions) == 1 and unique_versions[0] == version:
            # All apps have the same version as the main version
            prerelease_indicator = " 🧪 (prerelease)" if has_prerelease else ""
            summary.append(f"🏷️  **Version:** {version}{prerelease_indicator}")
        elif len(unique_versions) == 1:
            # All apps have the same version, but different from main version (increment mode)
            prerelease_indicator = " 🧪 (prerelease)" if has_prerelease else ""
            summary.append(f"🏷️  **Version:** {unique_versions[0]}{prerelease_indicator}")
        else:
            # Multiple different versions (mixed increment mode)
            summary.append("🏷️  **Versions:**")
            for item in matrix["include"]:
                app_version = item.get("version", version)
                prerelease_indicator = " 🧪" if is_prerelease_version(app_version) else ""
                summary.append(f"   - {item['app']}: {app_version}{prerelease_indicator}")
        
        summary.append("🛠️ **System:** Consolidated Release + OCI")
        
        if event_type == "workflow_dispatch":
            summary.append("📝 **Trigger:** Manual dispatch")
            if dry_run:
                summary.append("🧪 **Mode:** Dry run (no images published)")
        else:
            summary.append("📝 **Trigger:** Git tag push")
        
        summary.append("")
        summary.append("### 🐳 Container Images")
        if dry_run:
            summary.append("**Dry run mode - no images were published**")
        else:
            summary.append("Published to GitHub Container Registry:")
            # Get app metadata to determine correct image names
            all_apps = list_all_apps()
            app_domains = {app['name']: app['domain'] for app in all_apps}
            
            for item in matrix["include"]:
                app_name = item["app"]
                app_version = item.get("version", version)
                domain = app_domains.get(app_name, 'unknown')
                image_name = f"{domain}-{app_name}"
                summary.append(f"- `ghcr.io/{repository_owner.lower()}/{image_name}:{app_version}`")
        
        summary.append("")
        summary.append("### 🛠️ Local Development")
        summary.append("```bash")
        summary.append("# List all apps")
        summary.append("bazel run //tools:release -- list")
        summary.append("")
        summary.append("# Build and test an app locally")
        for app in apps[:2]:  # Show first 2 apps as examples
            summary.append(f"bazel run //tools:release -- build {app}")
        summary.append("```")
    
    return "\n".join(summary)

File: tools/kraken/summary.go
package kraken

import (
	"encoding/json"
	"fmt"
	"strings"
)

// GenerateReleaseSummary generates a release summary for GitHub Actions.
func GenerateReleaseSummary(matrixJSON, version, eventType string, dryRun bool, repositoryOwner string) string {
	var matrix MatrixConfig
	if matrixJSON != "" {
		if err := json.Unmarshal([]byte(matrixJSON), &matrix); err != nil {
			matrix = MatrixConfig{}
		}
	}

	var summary []string
	summary = append(summary, "## 🚀 Release Summary", "")

	if len(matrix.Include) == 0 {
		summary = append(summary, "🔍 **Result:** No apps detected for release")
		return strings.Join(summary, "\n")
	}

	summary = append(summary, "✅ **Result:** Release completed", "")

	var apps []string
	for _, item := range matrix.Include {
		apps = append(apps, item.App)
	}
	summary = append(summary, fmt.Sprintf("📦 **Apps:** %s", strings.Join(apps, ", ")))

	// Handle version display
	var appVersions []string
	for _, item := range matrix.Include {
		v := item.Version
		if v == "" {
			v = version
		}
		appVersions = append(appVersions, v)
	}

	uniqueVersions := uniqueStrings(appVersions)

	if len(uniqueVersions) == 1 && uniqueVersions[0] == version {
		summary = append(summary, fmt.Sprintf("🏷️  **Version:** %s", version))
	} else if len(uniqueVersions) == 1 {
		summary = append(summary, fmt.Sprintf("🏷️  **Version:** %s", uniqueVersions[0]))
	} else {
		summary = append(summary, "🏷️  **Versions:**")
		for _, item := range matrix.Include {
			v := item.Version
			if v == "" {
				v = version
			}
			summary = append(summary, fmt.Sprintf("   - %s: %s", item.App, v))
		}
	}

	summary = append(summary, "🛠️ **System:** Consolidated Release + OCI")

	if eventType == "workflow_dispatch" {
		summary = append(summary, "📝 **Trigger:** Manual dispatch")
		if dryRun {
			summary = append(summary, "🧪 **Mode:** Dry run (no images published)")
		}
	} else {
		summary = append(summary, "📝 **Trigger:** Git tag push")
	}

	summary = append(summary, "", "### 🐳 Container Images")
	if dryRun {
		summary = append(summary, "**Dry run mode - no images were published**")
	} else {
		summary = append(summary, "Published to GitHub Container Registry:")
		allApps, _ := ListAllApps()
		appDomains := make(map[string]string)
		for _, app := range allApps {
			appDomains[app.Name] = app.Domain
		}

		for _, item := range matrix.Include {
			appVersion := item.Version
			if appVersion == "" {
				appVersion = version
			}
			domain := appDomains[item.App]
			if domain == "" {
				domain = "unknown"
			}
			imageName := fmt.Sprintf("%s-%s", domain, item.App)
			summary = append(summary, fmt.Sprintf("- `ghcr.io/%s/%s:%s`", strings.ToLower(repositoryOwner), imageName, appVersion))
		}
	}

	summary = append(summary, "", "### 🛠️ Local Development", "```bash", "# List all apps", "bazel run //tools:release -- list", "")
	summary = append(summary, "# Build and test an app locally")
	limit := 2
	if len(apps) < limit {
		limit = len(apps)
	}
	for _, app := range apps[:limit] {
		summary = append(summary, fmt.Sprintf("bazel run //tools:release -- build %s", app))
	}
	summary = append(summary, "```")

	return strings.Join(summary, "\n")
}

func uniqueStrings(ss []string) []string {
	seen := make(map[string]bool)
	var result []string
	for _, s := range ss {
		if !seen[s] {
			seen[s] = true
			result = append(result, s)
		}
	}
	return result
}
